### DISCUSSION TITLE
datasets with variable maxzoom

### DISCUSSION THREAD

**Discussion by prusswan** [protomaps/PMTiles#321] [Q&A] [answered]
Related to https://github.com/syncpoint/openvtpk/issues/22, apparently vtpk can contain a mixture of maxzoom (15 and 16). So for the affected areas, the client needs to be able to retain data from the previous zoom, and not blank out just because no data exists at maxzoom.

https://github.com/protomaps/PMTiles/issues/48 seems to suggest Leaflet is not able to support such data, but MapLibre does, is this understanding correct?

**Comment by bdon** [comment]
(Moving to discussions)

the fixed maxzoom is a limitation of how extents are defined in Leaflet and MapLibre GL JS. There is a proposal in MapLibre GL JS that generated a lot of discussion for pyramids that are deeper in some areas than others, but I'm having trouble finding it right now. 

In any case PMTiles can store sparse pyramids, but how that variable overzoom is communicated to the client isn't well defined in any of those open source rendering implementations. If you can come up with such a definition that can be stored in the PMTiles metadata JSON, as it's not part of any existing spec like TileJSON.

**Reply by prusswan** [reply]
I need some time to digest this, but is the short answer simply to use such data with a lower effective maxzoom of 15? (some info loss but should be okay for most general usage) I don't recall running into such issues with raster tiles (perhaps the technology is just more forgiving) so still trying to wrap my head around why the older data (at lower zoom level) could not be expired in a more graceful manner.

**Reply by bdon** [reply]
The behavior in those renderers should be the same - if you load z15 and then scroll to z16, and the maxzoom is set to 16 but the tiles are missing / 404 responses, the z15 tile should remain over zoomed. However, this doesn't work in reverse: if you initialize the map to z16 it will not automatically traverse up the tile to find the closest matching zoom. Such logic would be difficult to implement without making wasteful requests or showing data at an inappropriate scale.

Yes, using a consistent z15 is the best solution for now.

**Reply by prusswan** [reply]
>  if you load z15 and then scroll to z16, and the maxzoom is set to 16 but the tiles are missing / 404 responses, the z15 tile should remain over zoomed. 

This might not be working properly (see below), even the tile inspector will discard z15 data when it is supposedly replaced by non-existent z16 data:

![image](https://github.com/protomaps/PMTiles/assets/671970/fe48c291-9548-480d-99c2-372835b7b1e0)


I have added some converted tiles (in both mbtiles/pmtiles) in case others are able to figure something out:

[OSM_firenze_vtpk.zip](https://github.com/protomaps/PMTiles/files/13883294/OSM_firenze_vtpk.zip)

**Reply by prusswan** [reply]
Surprisingly, only ol-pmtiles is able to work with such tiles properly (for overzooming). I'm not too familiar with maplibre's pmtiles protocol, but this seems to indicate a bug with tile handling.

**Reply by bdon** [reply]
Could It be that a Z15 tile is evicted once any of the 4 child Z16 tiles is successfully loaded, to avoid showing duplicated, overlapping tiles?

**Reply by prusswan** [reply]
Makes sense. Would it help to have a parameter to delay this eviction? For raster tiles, overlapping tiles is not an issue (since only whatever that is right on top would be visible), but for vector tiles, the parent tile could be subdivided and used to prefill the child tiles (even if they do not eventually load).

**Reply by bdon** [reply]
It might be a useful feature to propose in MapLibre but it's out of scope for the PMTiles library.

**Comment by etienneJr** [comment]
Hi,
I'm reopening the discussion because I'm having problems displaying sparse pmtiles in maplibre : as noticed above by @prusswan maplibre is not overzooming in for example z13 when z14 is not available.
Based on this [discussion on github/maplibre-gl-js](https://github.com/maplibre/maplibre-gl-js/discussions/5126), it seems to me that maplibre will overzoom if the requested tile returns a 404 error, whereas the pmtiles.js protocol response is empty without error when a tile is not available. Did I understand well ? How can I modify the pmtiles.pj protocol to test this hypothesis: returns a 404 error instead of a blank tile ?
Thanks !
E.

**Reply by prusswan** [reply]
At the moment, I know of these ways to use tilesets with uneven/variable maxzooms (without patching multiple client libraries to handle the different eviction scenarios):

1. split the tileset into multiple tilesets, each having a consistent maxzoom, e.g. (z1_to_z15, z16), then either:

A. use them together as multiple layers 

B. use tippecanoe to combine multiple tilesets with [--overzoom](https://github.com/felt/tippecanoe?tab=readme-ov-file#overzooming), however I'm not sure what will happen if there is collision between overzoomed tiles from one tileset and tiles from the other tileset, if they share the same tile coordinate (in X/Y/Z). Should not be too hard to fix though (maybe some logic to decide which tile to keep)

2. just limit the maxzoom to the global effective maxzoom (e.g. z15), without showing tiles from z16

**Reply by etienneJr** [reply]
Hi,
Thanks for your quick reply, I've taken some time to test the options.
> A. use them together as multiple layers

I don't like this option, because to apply the same style to both file sources, I have to duplicate the entire style definition (one with source=ID_file1, and exactly the same with source=ID_file2). By using styles with a lot of layers (e.g. OSM Bright or OSM Liberty), it makes a lot of duplicated content. And I'm really looking for a single-file solution that can be easily deployed by people with few map-coding skills.

> B. use tippecanoe to combine multiple tilesets with --overzoom

Thanks for the tip, it's a great feature that I didn't know about. here's an example of what I get at the limit of the detailed zone, seems perfect.
![image](https://github.com/user-attachments/assets/be14b1a6-7a36-4e5d-b7ad-1bbcae54107b)

But the calculation time is awfull: Merging a 1,7 Go z1-z12 file with a 50 Mo z13-z14 took 12h on my laptop. So my real use case (merging 10Go z1-z12 with  5Go z13-z14) may take ... weeks ?

And I noticed something else: in the detailed area, the overzoom of layer 12 mixes with layer 13 or 14. Almost everywhere, you can see straight roads (overzoomed from layer 12) crossing houses next to real curved roads (layer 13 or 14). So it seems that the overzoomed tile from z1-z12 file has been merge with the normal tile from z13-z14 file.
![image](https://github.com/user-attachments/assets/3d952e4a-2330-407b-9c70-e8487af5b828)

> 2. just limit the maxzoom to the global effective maxzoom (e.g. z15), without showing tiles from z16

I'm not sure I understand. If I limit the maxzoom to z12, my z13-z14 layers won't be visible, right ?

> without patching multiple client libraries

I wanted to patch pmtiles.js not to solve my specific issue, but to test whether it changed the behaviour of maplibre. Then I would have made (here or on maplibre) a request for additional functionality 😀​

**Reply by bdon** [reply]
You could try to throw an exception on this line https://github.com/protomaps/PMTiles/blob/main/js/src/adapters.ts#L254 instead of returning empty array buffer to simulate how getting a 404 might work with a ZXY tile source.

**Reply by etienneJr** [reply]
Thanks for the tip !! I've added this where indicated and it works perfectly !! (exactly the way I wanted 😀​)
```
const error = new Error(`tile ${z} ${x} ${y} not found`);
error.status = 404;
throw error;
```

For example here at the border, right side is z14 tile, and left side is overzoomed z12 tile (overzoomed by maplibre, not overzoomed by tipecanoe while generating the tiles)
![image](https://github.com/user-attachments/assets/63a794f9-2c4e-47a5-bffd-9fc17240a459)

The only drawback: a 404 error is displayed in the console for each tile not found. not visible to the user, but not very clean.

Now what ? How can we make this behaviour accessible to other people who would like to use sparse pmtiles?
- should I ask here for a new parameter in the pmtiles protocol to choose whether we want an empty tile or an error when no tile is available ? would that be possible?
- should I ask maplibre developpers for a parameter to ask for overzooming when the tile is empty ?

**Reply by bdon** [reply]
@HarelM is there any code path in maplibre GL that would let people take advantage of non-uniform maxzooms without relying on the error handling?

**Reply by HarelM** [reply]
We maybe able to carve one using a specific error type or something, but currently only 404 has that behavior.

**Reply by etienneJr** [reply]
I don't know if this is a useful information, but during my test, I observed the same behaviour (overzoom) as soon as I throw an error, whatever the error code 4xx (set with `error.status = 4xx`). Even values 5xx or 2xx gave the same behaviour, but maybe it is because I throw an Error ?

**Reply by bdon** [reply]
Yes, I think Maplibre is conservative and keeps the loaded tile if it encounters an error.

So what I see is that in some cases developers are depending on the maplibre error behavior for overzooming, and they cannot reproduce the functionality using PMTiles. I would prefer in the long term MapLibre has a real solution like defining tile pyramids, but as a temporary solution we can add a protocol parameter alongside `metadata` called `errorIfMissingTile`, that would let you opt into the maplibre behavior, what do you think?

**Reply by HarelM** [reply]
The following was suggested to solve this in maplibre from a specification point of view:
- https://github.com/maplibre/maplibre-style-spec/issues/938

Having said that, I think it would be much faster to implement something here as a temporary workaround until the above enhancement lands in Maplibre.
`errorIfMissingTile` sounds like a good name for this.

**Reply by bdon** [reply]
The error messages that would appear in the console if `errorIfMissingTIle` is still a problem but IMO an acceptable tradeoff because the same error messages happens with MapLibre and 404s. I think it is cleaner to patch this problem here instead of making a change to the maplibre specification in the above link. 

@etiennejourdier it is worth looking at the maplibre source code to see if there is any way to invoke this situation without the error message.

**Reply by etienneJr** [reply]
> it is worth looking at the maplibre source code to see if there is any way to invoke this situation without the error message.

Unfortunately I'm an amateur developer, so I clearly don't have the skills to delve into maplibre's source code...

**Reply by bdon** [reply]
Proposed PR here: https://github.com/protomaps/PMTiles/pull/505

can you provide a minimal real world sample of a sparse tileset?

**Reply by prusswan** [reply]
> Proposed PR here: #505
> 
> can you provide a minimal real world sample of a sparse tileset?

I attached some sample previously. It was extracted from Esri's OSM vector layer as vtpk (which led to this issue)

@etiennejourdier I think there are different situations, so some of the alternatives may be better if you have no control over the tiles or the client code, but I understand the issue ultimately lies with the specs, and libraries are not obligated to support it (since there is no "standard" to follow). For tile generation, keeping the tilesets/layers separate is more flexible (since they can be merged anytime, for simple cases with no overlaps at least), but having more options is always good.

**Reply by bdon** [reply]
![CleanShot 2024-12-15 at 16 40 42@2x](https://github.com/user-attachments/assets/faa27425-baea-4da9-92c8-a1ff78c85c20)

Attached is screenshot of how #505 works with the sample file - it shows an error in the console but simultaneously displays z14 and z15 tiles.

**Reply by bdon** [reply]
@etiennejourdier @prusswan I published pmtiles v4.1.0 with an `errorOnMissingTile` option:
```js
let protocol = new pmtiles.Protocol({metadata: true, errorOnMissingTile: true});
```

**Reply by etienneJr** [reply, marked as answer]
That's perfect, thanks a lot for your responsiveness !! I have just deployed this on my website (a playground directory). You can see [here](https://playguide.eu/app/playground/1738f899-d174-4be0-9987-03231b60645a) the behaviour at the border between the z12 zone (north) and the z14 zone (south).

**Comment by HarelM** [comment]
> There is a proposal in MapLibre GL JS that generated a lot of discussion for pyramids that are deeper in some areas than others, but I'm having trouble finding it right now.

Here it is for future reference:
- https://github.com/maplibre/maplibre-style-spec/issues/60
